### DISCUSSION TITLE
Extended sorting options for .dbc serialisation

### DISCUSSION THREAD

**Discussion by mon** [cantools/cantools#452] [Ideas]
I have a fairly large .dbc file that I want to auto-format for consistency. I'm already a fair way there by just loading and then re-saving the .dbc with my preferred `sort_signals` function.

However, I also want to apply much more specific formatting to other sections
- attributes sorted by node attributes, then messages, then by attribute name, message ID, and signal name
- comments sorted by message ID, in order of BU/BO/SG, and signal name
- value table defs sorted by message ID then signal name
- value tables themselves sorted by value
- no 32 character limit on signal names

I'm currently doing all this by
- Hooking `cantools.database.can.formats.dbc._dump_[attributes|comments|choices]` and sorting the resulting string
- Overriding `Signal.choices` with my own sorted one
- Overriding `database.can.formats.dbc.LongNamesConverter.convert` with one that always returns `None`

I'm not very comfortable with the long-term stability of poking around in the internals like this, except maybe the `Signal.choices` replacement since that's public.

Would it be too much to open a PR that added `sort_attributes`, `sort_comments`, `sort_choices` functions to tune the sorting behaviour? Or are there better ways to do this? I don't want to go to the effort to discover it was a poor design choice.

**Comment by juleq** [comment]
Hi @mon,

I think adding parameters that default to the previous behavior is fine and adding more optional control with respect to sorting is a sensible addition.

Please add tests to ensure your additions remain intact.

@andlaus agreed?

**Reply by andlaus** [reply]
agreed

**Reply by mon** [reply]
Alright! I'll get to it 👍

**Reply by mon** [reply]
Some initial questions
- Why is the default for sort_signals `'default'` instead of None? It gets coerced to None for the internal string dumper anyway, and might make my changes cleaner to have an `Optional[callback]` instead of `Union[callback, Literal['default']]`
- All of the `as_xxx_string` functions check `self._sort_signals` but that value isn't used. Fine to remove?
- Do you have a preferred code auto-formatter or just match the existing style?

**Reply by erzoe** [reply]
`sort_signals = None` means no sorting, `sort_signals = SORT_SIGNALS_DEFAULT` means the default sorting (which is different for different file types). I cannot say why the defaults have been chosen the way they were, I have only kept backward compatibility when adding the `sort_signals` parameter.

The `sort_signals` argument passed to the `Database` constructor is passed through to `load_string` when reading a file.

When saving the database the `sort_signals` argument which has been passed to constructor is usually ignored and the `sort_signals` argument passed to the `as_xxx_string` method is used instead.
The exception is if `sort_signals = None` has been passed to the constructor then SORT_SIGNALS_DEFAULT is changed to None in `as_xxx_string`. I have added that for convenience because when I am telling the constructor that I don't want automatic sorting I would not expect that I need to repeat myself when saving the file.

Does that answer your first two questions?

**Reply by mon** [reply]
That clarifies everything, thanks for the info!

**Comment by andlaus** [comment]
Personally, I'm fine with your proposals, but I remember that there was a reason for the selection mechanism, I just don't remember what it was. if I didn't make a mistake @erzoe implemented this in #382, maybe he knows more...

regarding the coding style, we're somewhat relaxed (i.e., 100% consistency or usage of an auto formatter is not necessary). Just try to be in line with the existing code and use common sense (but please don't mix style- and functional changes in the same pull request)...

**Reply by mon** [reply]
Excellent, thanks! This will make future PRs easier too :)

**Comment by mon** [comment]
In case Google indexes this thread, these changes have been merged as part of #456 !
